vowpalwabbit/exploration.cc

```diff
@@ -0,0 +1,121 @@
+#include "exploration.h"
+
+#include <stdexcept>
+#include <algorithm>
+#include <numeric>
+#include <cstring>
+
+using namespace std;
+
+namespace exploration
+{
+  void epsilon_greedy(float epsilon, uint32_t top_action, float* probability_distribution, uint32_t num_actions)
+  {
+    if (num_actions == 0)
+      return;
+
+    if (top_action >= num_actions)
+      throw out_of_range("top_action must be smaller than num_actions");
+
+    float prob = epsilon / (float)num_actions;
+
+    // size & initialize vector to prob
+    float* prob_end = probability_distribution + num_actions;
+    for (float* d = probability_distribution; d != prob_end; ++d)
+      *d = prob;
+
+    probability_distribution[top_action] += 1.f - epsilon;
+  }
+
+  void softmax(float lambda, const float* scores, float* probability_distribution, uint32_t num_actions)
+  {
+    if (num_actions == 0)
+      return;
+
+    float norm = 0.;
+    float max_score = *std::max_element(scores, scores + num_actions);
+
+    float* prob_end = probability_distribution + num_actions;
+    const float *s = scores;
+    for (float *d = probability_distribution; d != prob_end; ++d, ++s)
+    {
+      float prob = exp(lambda*(*s - max_score));
+      norm += prob;
+
+      *d = prob;
+    }
+
+    // normalize
+    if (norm > 0)
```

Comment:
This if seems strange, because there is no 'else'?  

Is the 'if' even required?  exp necessarily returns a number >= 0 and one number (the max_score number) is guaranteed to be > 0.

Reply:
done. I think this was left over from previous code that didn't check if there are any scores...